daal4py/sklearn/neighbors/knn.py

```diff
@@ -0,0 +1,206 @@
+
+            prediction_alg = knn_classification_prediction(**alg_params)
+
+            X = d4p.get_data(X)
+            prediction_result = prediction_alg.compute(X, training_result.model)
+
+            if return_distance:
+                results = prediction_result.distances.astype(fptype), prediction_result.indices
+            else:
+                results = prediction_result.indices
+        else:
+            return super(KNeighborsMixin, self).kneighbors(X, n_neighbors, return_distance)
+
+        if chunked_results is not None:
+            if return_distance:
+                neigh_dist, neigh_ind = zip(*chunked_results)
+                results = np.vstack(neigh_dist), np.vstack(neigh_ind)
+            else:
+                results = np.vstack(chunked_results)
+
+        if not query_is_train:
+            return results
+        else:
+            # If the query data is the same as the indexed data, we would like
+            # to ignore the first nearest neighbor of every sample, i.e
+            # the sample itself.
+            if return_distance:
+                neigh_dist, neigh_ind = results
+            else:
+                neigh_ind = results
+
+            n_queries, _ = X.shape
+            sample_range = np.arange(n_queries)[:, None]
+            sample_mask = neigh_ind != sample_range
+
+            # Corner case: When the number of duplicates are more
+            # than the number of neighbors, the first NN will not
+            # be the sample, but a duplicate.
+            # In that case mask the first duplicate.
+            dup_gr_nbrs = np.all(sample_mask, axis=1)
+            sample_mask[:, 0][dup_gr_nbrs] = False
+            neigh_ind = np.reshape(
+                neigh_ind[sample_mask], (n_queries, n_neighbors - 1))
+
+            if return_distance:
+                neigh_dist = np.reshape(
+                    neigh_dist[sample_mask], (n_queries, n_neighbors - 1))
+                return neigh_dist, neigh_ind
+            return neigh_ind
+
+
+class KNeighborsClassifier(BaseKNeighborsClassifier, KNeighborsMixin):
+    def predict(self, X):
+        X = check_array(X, accept_sparse='csr')
+
+        try:
+            fptype = getFPType(X)
+        except ValueError:
+            fptype = None
+
+        if daal_check_version((2020, 3), (2021, 9)) \
+        and self.weights in ['uniform', 'distance'] and self.algorithm in ['brute', 'kd_tree'] \
+        and (self.metric == 'minkowski' and self.p == 2 or self.metric == 'euclidean') \
+        and self._y.ndim == 1 and fptype is not None and not sp.issparse(X):
+
+            n_classes = len(self.classes_)
+
+            if self.algorithm == 'brute':
+                knn_classification_training = d4p.bf_knn_classification_training
+                knn_classification_prediction = d4p.bf_knn_classification_prediction
+                # Brute force method always computes in doubles due to precision need
+                compute_fptype = 'double'
```

Comment:
float data be converted inside the algorithm to double type? How does sklearn get around this?

Reply:
sklearn converts data to double in blocks